File: frontend/src/services/authService.ts
import { apiClient } from "./apiClient";
import type { AuthToken, LoginRequest, RegisterRequest, SetupStatus, User } from "../types/auth";

export const authService = {
  async getSetupStatus(): Promise<SetupStatus> {
    const response = await apiClient.get<SetupStatus>("/api/auth/setup-status");
    return response.data;
  },

  async register(payload: RegisterRequest): Promise<User> {
    const response = await apiClient.post<User>("/api/auth/register", payload);
    return response.data;
  },

  async login(payload: LoginRequest): Promise<AuthToken> {
    const response = await apiClient.post<AuthToken>("/api/auth/login", payload);
    return response.data;
  },

  async me(): Promise<User> {
    const response = await apiClient.get<User>("/api/auth/me");
    return response.data;
  }
};

File: backend/app/services/auth_service.py
from fastapi import HTTPException, status
from sqlalchemy.exc import IntegrityError
from sqlalchemy.orm import Session

from app.core.security import create_access_token, hash_password, verify_password
from app.repositories.user_repository import UserRepository
from app.schemas.auth import LoginRequest, RegisterRequest, SetupStatus, Token


class AuthService:
    def __init__(self, db: Session):
        self.db = db
        self.users = UserRepository(db)

    def setup_status(self) -> SetupStatus:
        user_count = self.users.count()
        return SetupStatus(needs_owner=user_count == 0, user_count=user_count)

    def register(self, payload: RegisterRequest):
        if self.users.get_by_email(payload.email):
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail="A user with this email already exists.",
            )

        user = self.users.create(
            name=payload.name,
            email=payload.email,
            password_hash=hash_password(payload.password),
            avatar_url=payload.avatar_url,
        )
        try:
            self.db.commit()
        except IntegrityError as exc:
            self.db.rollback()
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail="A user with this email already exists.",
            ) from exc
        self.db.refresh(user)
        return user

    def login(self, payload: LoginRequest) -> Token:
        user = self.users.get_by_email(payload.email)
        if not user or not verify_password(payload.password, user.password_hash):
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Incorrect email or password.",
            )
        if not user.is_active:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="This account is inactive.",
            )

        return Token(access_token=create_access_token(str(user.id)))
